### PR TITLE
Use message header flag to distinguish secure session control msgs

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
We are using two bytes in every message (including data packets) to identify secure session control message type. The field is applicable to only control messages. Also, the secure session control message type should be in the encrypted section of the message.

 #### Summary of Changes
Reserve one bit in the `flags` to distinguish secure session control message from data message. The actual message type will be moved into encrypted section.

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

**Comment by woody-apple** [comment]
@jelderton @hawk248 @andy31415 ?

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
We seem to incrementally update the protocol. Any chance we could create a doc/spec with the protocol we intend to implement?

For the header, https://openweave.io/documents/protocol-specification-weave-message-layer.pdf was used as a baseline except we removed tunnelling and other encryption flags.

It is probably worth it to get one pass over things and find out which parts we want and which parts we do not want to keep and try to minimize intermediate versions as they are incompatible with each other.

**Review by gerickson — CHANGES_REQUESTED** [review]
While I have no concrete, quantitative quibbles with this pull request in isolation; I find it impossible to review this and other secure-channel-related pull requests in the absence of a one-pager or specification because I cannot correlate with what's in the pull request against said document or roadmap that details where things are headed.

Consequently, before approving this pull request, can someone or a team produce such a document and/or road map?

**Comment by balducci-apple** [comment]
@gerickson @andy31415 The secure channel tiger team will try and get a document to yalls ASAP.

**Review by hawk248 — APPROVED** [review]
👍

**Review by emargolis — APPROVED** [review]

File: src/transport/MessageHeader.cpp
Comment (emargolis) [inline src/transport/MessageHeader.cpp:35]: Just because we don't have a spec that describes this I am going to ask here:
 - What about messages that are not encrypted and the session doesn't exist?
 - What about messages that are encrypted using Group Keys?
 - Can we use 16-bit identifier?

Just for the reference in OpenWeave we have an equivalent field "KeyID":
 - 16-bit value
 - KeyID = 0 means that the message is not encrypted
 - 4-bits of the value are used to encode the type of the key: session key, type of a group key...

File: src/transport/MessageHeader.cpp
Comment (emargolis) [inline src/transport/MessageHeader.cpp:36]: What about support for tags that are 12-bytes and 16-bytes?

Comment (pan-apple) [reply src/transport/MessageHeader.cpp:36]: Do we need support for tags of 12/16 bytes in CHIP? If so, need to update the structure to accommodate those tags.

File: src/transport/MessageHeader.cpp
Comment (pan-apple) [reply src/transport/MessageHeader.cpp:35]: @balducci-apple what are your thoughts about this? We can probably address these in the spec document.

File: src/transport/MessageHeader.cpp
Comment (emargolis) [reply src/transport/MessageHeader.cpp:36]: Yes, CHIP supports 3 tag lengths for CCM: 8, 12, and 16
The idea is that tag length is negotiated once during session establishment and only this tag length is used within context of that session.
Few bits of the 16-bit header can be allocated to encode the tag length. However, I might not be necessary because SessionID will identify that only specific tag length can be used for messages sent over that session.

Comment (pan-apple) [reply src/transport/MessageHeader.cpp:36]: Sounds good. Agreed that SessionID can be used to identify what the tag length.

**Comment by yufengwangca** [comment]
> While I have no concrete, quantitative quibbles with this pull request in isolation; I find it impossible to review this and other secure-channel-related pull requests in the absence of a one-pager or specification because I cannot correlate with what's in the pull request against said document or roadmap that details where things are headed.
> 
> Consequently, before approving this pull request, can someone or a team produce such a document and/or road map?

I second Grant, we need to first agree on the Header format spec before land this PR.

https://github.com/project-chip/connectedhomeip/issues/1692 is tracking this effort, we use Weave Message Format specification (https://openweave.io/documents/protocol-specification-weave-message-layer.pdf) as a baseline except we removed tunneling part.

The following Header format definition represents our latest progress.

Header fields that are not encrypted:
16 bit: | VERSION: 4 bit | FLAGS: 4 bit | Encrypt Type: 4 bit | RESERVED: 4 bit
32 bit: | MESSAGE_ID
64 bit: | SOURCE_NODE_ID (if source node flag is set)
64 bit: | DEST_NODE_ID (if destination node flag is set)
16 bit: | Key ID
16 bit: | Payload Length
Variable: | IV

Header fields that are encrypted:
8 bit: | Exchange Header
8 bit: | Exchange Message Type // (formally 16-bit Secure message type)
16 bit: | Exchange ID // (formally 32-bit Secure Session ID)
32 bit: | Message Profile ID
Variable: | Application data

Integrity checksum are not encrypted:
Variable: | Message Authentication Tag

**Review by yufengwangca — CHANGES_REQUESTED** [review]
Align the header format implementation based on the header format we agreed in
https://github.com/project-chip/connectedhomeip/issues/1692

**Comment by pan-apple** [comment]
Closing this PR, as the suggested specifications require more comprehensive changes to the header format.

File: src/transport/MessageHeader.cpp
Comment (turon) [inline src/transport/MessageHeader.cpp:58]: What is the proposed enumeration for SecureSessionControlMessage types?

File: src/transport/MessageHeader.h
Comment (turon) [inline src/transport/MessageHeader.h:131]: Why limit the scope of message types to only session control?

**Review by gerickson — APPROVED** [review]
Approving after offline discussion with @robszewczyk.

**Pull request closed without merging**
